Source/CarthageKit/VersionFile.swift

```diff
@@ -0,0 +1,270 @@
+		return curry(self.init)
+			<^> j <| VersionFile.commitishKey
+			<*> j <| VersionFile.xcodeVersionKey
+			<*> j <||? Platform.macOS.rawValue
+			<*> j <||? Platform.iOS.rawValue
+			<*> j <||? Platform.watchOS.rawValue
+			<*> j <||? Platform.tvOS.rawValue
+	}
+}
+
+/// Creates a version file for the current dependency in the
+/// Carthage/Build directory which associates its commitish with
+/// the SHA1s of the built frameworks for each platform in order
+/// to allow those frameworks to be skipped in future builds.
+///
+/// Returns a signal that succeeds once the file has been created.
+public func createVersionFileForDependency(_ dependency: Dependency<PinnedVersion>, forPlatforms platforms: Set<Platform>, buildProductURLs: [URL], rootDirectoryURL: URL) -> SignalProducer<(), CarthageError> {
+	var platformCaches: [String: [CachedFramework]] = [:]
+
+	let platformsToCache = platforms.isEmpty ? Set(Platform.supportedPlatforms) : platforms
+	for platform in platformsToCache {
+		platformCaches[platform.rawValue] = []
+	}
+
+
+	let writeVersionFile = currentXcodeVersion()
+		.attemptMap { xcodeVersion -> Result<(), CarthageError> in
+			let rootBinariesURL = rootDirectoryURL.appendingPathComponent(CarthageBinariesFolderPath, isDirectory: true).resolvingSymlinksInPath()
+			let versionFileURL = rootBinariesURL.appendingPathComponent(".\(dependency.project.name).version")
+
+			let versionFile = VersionFile(
+				commitish: dependency.version.commitish,
+				xcodeVersion: xcodeVersion,
+				macOS: platformCaches[Platform.macOS.rawValue],
+				iOS: platformCaches[Platform.iOS.rawValue],
+				watchOS: platformCaches[Platform.watchOS.rawValue],
+				tvOS: platformCaches[Platform.tvOS.rawValue])
+
+			return versionFile.write(to: versionFileURL)
+	}
+
+	if !buildProductURLs.isEmpty {
+		return SignalProducer<URL, CarthageError>(buildProductURLs)
+		.flatMap(.merge) { url -> SignalProducer<(), CarthageError> in
+			let platformName = url.deletingLastPathComponent().lastPathComponent
+			let frameworkName = url.deletingPathExtension().lastPathComponent
+			let frameworkURL = url.appendingPathComponent(frameworkName, isDirectory: false)
+			return md5ForFileAtURL(frameworkURL)
+				.attemptMap { md5 -> Result<(), CarthageError> in
+					let cachedFramework = CachedFramework(name: frameworkName, md5: md5)
+					if var frameworks = platformCaches[platformName] {
+						frameworks.append(cachedFramework)
+						platformCaches[platformName] = frameworks
+						return .success(())
+					}
+					else {
+						return .failure(.versionFileError(description: "unexpected platform found in path"))
+					}
+				}
+				.collect()
+				.flatMap(.concat) { _ -> SignalProducer<(), CarthageError> in
+					return writeVersionFile
+				}
+		}
+	} else {
+		// Write out an empty version file for dependencies with no built frameworks, so cache builds can differentiate between
+		// no cache and a dependency that has no frameworks
+		return writeVersionFile
```

Comment:
It's important because the commitish still needs to be validated to determine if downstream dependencies should be rebuilt. For example, if you have an xcodeConfigs repo that changes it's version that dependency B relies on, then you'll want to recompile B even if B is still the same version that was last compiled. 

Writing out an empty file distinguishes between "We ran a build, and determined that this dependency had no frameworks" and "This dependency didn't exist / wasn't considered during the last build, so rebuild"